modules/core/src/main/scala/gem/Location.scala

```diff
@@ -10,46 +11,167 @@ import scalaz.Ordering.{EQ, GT, LT}
-  // agree.
+  // These functions aren't of any use to clients.  Instead they are involved
+  // in the cacluation of Locations that fall between two other locations.
 
-  def toList: List[Int] =
-    toIList.toList
+  /** Infinite Stream of position elements corresponding to this Location. */
+  protected def positions: Stream[Int]
 
-  override def toString: String =
-    toList.mkString("{", ",", "}")
+  /** Minimum size required to obtain all the fixed elements in this Location,
+    * if any.  Beginning and End Locations have no fixed elements.
+    */
+  protected def minPrefixLength: Int
 }
 
 object Location {
-  private def trim(is: Seq[Int]): IList[Int] =
-    trim(IList(is: _*))
 
-  // Remove trailing Int.MinValue so that == agrees with EQ.
-  private def trim(is: IList[Int]): IList[Int] =
-    is.dropRightWhile(_ == Int.MinValue)
+  /** A marker for the first Location.  No other Location comes before the
+    * Beginning.  Beginning should not be used to order elements unless you
+    * are sure no elements will ever need to be inserted before Beginning.  Use
+    * Beginning to calculate Locations that fall before the first existing
+    * Location (if any).
+    */
+  case object Beginning extends Location {
+    protected def positions: Stream[Int] =
+      Stream.continually(Int.MinValue)
+
+    protected def minPrefixLength: Int =
+      0
+
+    override def toString: String =
+      "{α}"
+  }
+
+  /** A Location that falls somewhere between the Beginning and End.
+    */
+  sealed abstract case class Middle(posList: OneAnd[IList, Int]) extends Location {
+    def positions: Stream[Int] =
+      posList.toStream #::: Stream.continually(Int.MinValue)
+
+    def toIList: IList[Int] =
+      posList.head +: posList.tail
+
+    def toList: List[Int] =
+      posList.head :: posList.tail.toList
+
+    protected def minPrefixLength: Int =
+      1 + posList.tail.length
+
+    override def toString =
+      toList.mkString("{", ",", "}")
+  }
+
+  /** A marker for the last Location.  No other Location comes after the End.
+    * End should not be used to order elements unless you are sure no elements
+    * will ever need to be inserted after End.  Use End to calculate new
+    * Locations that fall after the last existing Location (if any).
+    */
+  case object End extends Location {
+    def positions: Stream[Int] =
+      Stream.continually(Int.MaxValue)
+
+    protected def minPrefixLength: Int =
+      0
+
+    override def toString: String =
+      "{ω}"
+  }
+
+  // Constructors
 
   def apply(is: Int*): Location =
-    new Location(trim(is)) {}
+    fromTrimmedIList(IList(is: _*))
+
+  def fromFoldable[F[_]: Foldable](fi: F[Int]): Location =
+    fromTrimmedIList(IList.fromFoldable(fi))
+
+  private def fromTrimmedIList(is: IList[Int]): Location =
+    is.dropRightWhile(_ === Int.MinValue) match {
+      case INil()      => Beginning
+      case ICons(h, t) => new Location.Middle(OneAnd(h, t)) {}
```

Comment:
To prevent anybody from directly using the constructor.  I'm trying to ensure that `==` and `===` provide the same answer in all cases.  See the definition of `OrderLocation` to see why that might not be the case otherwise.